feat: add zprint formatter for clojure

agriffis opened the pull request:
This pull request adds a formatter for Clojure using zprint.

zprint unfortunately lacks the ability to format a range, so we implement `:FormatLines` by sending only the lines to be formatted. I've opened a PR to add range formatting to zprint, see https://github.com/kkinnear/zprint/issues/122

Doc is updated by vimdoc, with a special section `*codefmt-mappings-zprint*` suggesting an alternative mapping for top-level forms which depends on also having [vim-sexp](https://github.com/guns/vim-sexp) installed:

    autocmd FileType clojure nmap <buffer> <silent> <leader>== <leader>=iF

dbarnett commented on .travis.yml:
```diff
@@ -20,6 +22,9 @@ before_script:
   - git clone https://github.com/google/vim-glaive.git ../glaive/
+  - wget https://github.com/kkinnear/zprint/releases/download/${ZPRINT_VERSION}/zprintl-${ZPRINT_VERSION}
+  - sudo mv zprintl-${ZPRINT_VERSION} /usr/local/bin/zprint
+  - sudo chmod +x /usr/local/bin/zprint
```
I think the GNU `install` command is supposed to move and chmod files in one go. Can you check if that will work here?

agriffis replied:
I had considered that, but GNU `install` copies rather than moves and that seemed messy. I don't have strong feelings about it—just let me know what you'd like.

agriffis replied:
Done

dbarnett commented on autoload/codefmt/zprint.vim:
```diff
@@ -0,0 +1,82 @@
+
+let s:plugin = maktaba#plugin#Get('codefmt')
+
+let s:supported_filetypes = ['clojure']
```
Style nit: Is this meant to be constant or updateable? If constant, consider capitalizing and doing `lockvar!` since vimscript functions make it easy to accidentally modify lists otherwise (and you'll probably want an `if exists()` guard since this is an autoload file). Otherwise, you could attach it to `s:plugin.globals`.

agriffis replied:
That's from `autoload/codefmt/prettier.vim` but it would make more sense to use the simpler `AppliesToBuffer` function from `autoload/codefmt/rustfmt.vim` since there's just one filetype.

agriffis replied:
Done

dbarnett commented on autoload/codefmt/zprint.vim:
```diff
@@ -0,0 +1,82 @@
+    call maktaba#ensure#IsNumber(a:endline)
+    let l:lines = getline(1, line('$'))
+
+    " zprint doesn't support formatting a range of lines, so format the range
```
Can you follow up with zprint to request real range formatting and post a pointer to an issue in their tracker here? For a functional language, hacked range formatting is going to be especially brittle.

agriffis replied:
Will do

agriffis replied:
Done, but I'll also update the PR description to include this link before removing the draft status. https://github.com/kkinnear/zprint/issues/122

dbarnett commented on vroom/zprint.vroom:
```diff
@@ -0,0 +1,95 @@
+  (defn x [] (cond nil 1 :else 2))
+  @end
+
+It is also the default formatter for the clojure file type.
```
The descriptions would read better if you reordered the sections so this `It is also the default formatter for` section is right after the `You can format any buffer` section.

agriffis replied:
Will do

agriffis replied:
I didn't reorder exactly like that, but I reworked the narrative so it's clearer and doesn't have a continuity dependency between the sections. (I wanted to keep `:FormatCode` and `:FormatLines` next to each other, but previously the test for `:FormatCode` was worded to emphasize calling zprint explicitly.)

agriffis commented on instant/flags.vim:
```diff
@@ -115,3 +115,18 @@ call s:plugin.Flag('prettier_options', [
 call s:plugin.Flag('prettier_executable', 'prettier')
+
+""
+" Whether and where to change directory before calling zprint. If this is true,
```
Yeah, I got rid of this option in a later commit, with the same thinking. I was planning to squash the commits when I remove the draft status on the PR

agriffis commented on vroom/zprint.vroom:
```diff
@@ -0,0 +1,81 @@
+
+  % (defn x [] (cond nil 1 :else 2))
+  :FormatCode zprint
+  ! zprint .*
```
Thanks, I eventually figured this out too. Tests are passing now.

agriffis commented:
@dbarnett Thanks for the review. I've squashed the commits and this should be ready to merge unless you have any other concerns.

agriffis commented:
If you pull #131 before this, then I'll rebase this PR to resolve the minor merge conflict

agriffis commented:
Rebased to resolve merge conflict with #131

dbarnett commented on .travis.yml:
```diff
@@ -23,6 +24,8 @@ before_script:
   - sudo dpkg -i ./vroom_${VROOM_VERSION}-1_all.deb
   - git clone -b ${MAKTABA_VERSION} https://github.com/google/vim-maktaba.git ../maktaba/
   - git clone https://github.com/google/vim-glaive.git ../glaive/
+  - wget https://github.com/kkinnear/zprint/releases/download/${ZPRINT_VERSION}/zprintl-${ZPRINT_VERSION}
```
Wait, do you actually need zprint installed for these if you're faking the I/O?

agriffis replied:
I totally did not understand that aspect of vroom. I think the tests will need another pass to straighten this out, but then I think you're right.

agriffis replied:
@dbarnett Rebased now on master with updated tests to fake I/O properly, and removed zprint install from travis.

agriffis commented:
@dbarnett In case this was lost in code review comments, I updated the tests to use fake I/O fully, and removed the zprint install from travis. Let me know if there's anything else...

dbarnett commented on autoload/codefmt/zprint.vim:
```diff
@@ -0,0 +1,91 @@
+
+""
+" @section Recommended zprint mappings, mappings-zprint
+" @parentsection mappings
```
Note to self: Build a new vimdoc release. I just found that the `@parentsection` directive isn't supported yet in the latest vimdoc 0.6.0 release.
